src/payment/strategies/zip/zip-payment-strategy.ts

```diff
@@ -0,0 +1,159 @@
+    }
+
+    finalize(options?: PaymentRequestOptions): Promise<InternalCheckoutSelectors> {
+        return Promise.reject(new OrderFinalizationNotRequiredError());
+    }
+
+    private _displayZip(payload: OrderRequestBody, options?: PaymentRequestOptions): Promise<InternalCheckoutSelectors> {
+        return this._store.dispatch(this._paymentStrategyActionCreator.widgetInteraction(() => {
+            this._store.dispatch(this._paymentMethodActionCreator.loadPaymentMethod(this._methodId))
+                .then(state => {
+                    if (!this._zipClient) {
+                        throw new NotInitializedError(NotInitializedErrorType.PaymentNotInitialized);
+                    }
+
+                    this._zipClient.Checkout.init(this._getZipPayload());
+                });
+
+            // Lightbox closing event listener
+            return new Promise((resolve, reject) => {
+                this._lightboxEvents$
+                    .pipe(take(1))
+                    .subscribe((event: { type: ZipModalEvent }) => {
+                        if (event.type === ZipModalEvent.CancelCheckout) {
+                            reject(new PaymentMethodCancelledError());
+                        }
+                        if (event.type === ZipModalEvent.CheckoutApproved) {
+                            resolve(this._submitOrder(payload, options));
+                        }
+                    });
+            });
+        }, { methodId: this._methodId }), { queueId: 'widgetInteraction' });
+    }
+
+    private _getZipPayload(): ZipPayload {
+        return {
+            onComplete: this._handleResponse,
+            onCheckout: this._fetchPayload,
+        };
+    }
+
+    private _submitOrder(payload: OrderRequestBody, options?: PaymentRequestOptions): Promise<InternalCheckoutSelectors> {
+        const order = { useStoreCredit: payload.useStoreCredit };
+
+        if (!this._zipLightboxResponse) {
+            throw new MissingDataError(MissingDataErrorType.MissingPaymentMethod);
+        }
+
+        const paymentPayload = {
+            methodId: this._methodId,
+            paymentData: { nonce: this._zipLightboxResponse.checkoutId } as NonceInstrument,
+        };
+
+        return this._store.dispatch(this._orderActionCreator.submitOrder(order, options))
+            .then(() => this._store.dispatch(this._paymentActionCreator.submitPayment(paymentPayload)));
+    }
+
+    @bind
+    private _fetchPayload(resolve: (response: ZipPostResponse) => void): void {
+        this._store.dispatch(this._paymentMethodActionCreator.loadPaymentMethod(this._methodId))
+            .then(state => {
+                const paymentMethod = state.paymentMethods.getPaymentMethod(this._methodId);
+
+                if (!paymentMethod || !paymentMethod.clientToken) {
+                    throw new MissingDataError(MissingDataErrorType.MissingPaymentMethod);
+                }
+
+                resolve(JSON.parse(paymentMethod.clientToken));
```

Comment:
We are returning a JSON parseable string because, the zip token creation return 2 values a checkout id and a URI, we need these 2 values to can display the lightbox

Reply:
I understand that Zip might have a special requirement. But it just feels like a shortcut / snowflake to me as `clientToken` field isn't supposed to contain a JSON parsable string. Anyway, I'll defer the decision to the payments team. cc @bc-anna @IamDavidovich.